rtrlib/pfx/trie/trie-pfx.c

```diff
@@ -83,34 +83,54 @@ void pfx_table_free(struct pfx_table *pfx_table)
 
 int pfx_table_append_elem(struct node_data *data, const struct pfx_record *record)
 {
-    data->len++;
-    data->ary = realloc(data->ary, sizeof(struct data_elem) * data->len);
-    if(data->ary == NULL)
+    struct data_elem *tmp  = realloc(data->ary, sizeof(struct data_elem) * ((data->len) + 1));
+    if(tmp == NULL)
         return PFX_ERROR;
+    data->len++;
+    data->ary = tmp;
     data->ary[data->len - 1].asn = record->asn;
     data->ary[data->len - 1].max_len = record->max_len;
     data->ary[data->len - 1].socket = record->socket;
     return PFX_SUCCESS;
 }
 
-int pfx_table_create_node(struct trie_node **node, const struct pfx_record *record)
+int pfx_table_create_node(struct trie_node **node,
+			  const struct pfx_record *record)
 {
+    int err;
+
     *node = malloc(sizeof(struct trie_node));
     if(*node == NULL)
         return PFX_ERROR;
+
     (*node)->prefix = record->prefix;
     (*node)->len = record->min_len;
     (*node)->lchild = NULL;
     (*node)->rchild = NULL;
     (*node)->parent = NULL;
 
     (*node)->data = malloc(sizeof(struct node_data));
-    if((*node)->data == NULL)
-        return PFX_ERROR;
+    if((*node)->data == NULL) {
+			err = PFX_ERROR;
+			goto free_node;
+    }
+
     ((struct node_data *) (*node)->data)->len = 0;
     ((struct node_data *) (*node)->data)->ary = NULL;
 
-    return pfx_table_append_elem(((struct node_data *) (*node)->data), record);
+    err = pfx_table_append_elem(((struct node_data *) (*node)->data), record);
+    if (err)
+	    goto free_node_data;
+
+    return PFX_SUCCESS;
+
+
+free_node_data:
```

Comment:
I agree, iff there are lots of error cases (N > 2, 😉) that have to do similar cleanup operations before return. However, here we have just 2 error cases; I admit, one is reusing (or requires) the same free() as the other. 

IMHO this could be written nicely with 2 nested`if () else`. However, you may leave as is - just my personal view.